app/validation/validator.py

```diff
@@ -506,6 +507,59 @@ def _validate_routing_rule(self, rule, answer_ids_with_group_id, block_or_group)
 
         return errors
 
+    def _validate_answer_routing_value(self, rule, answer_ids_with_parent_id):
+        errors = []
+
+        when_rule = rule.get('goto').get('when')
+        if when_rule:
+            for when_dict in when_rule:
+                amount_matched = 0
+                errors.extend(self._validate_routing_when_dict(when_dict, answer_ids_with_parent_id, amount_matched))
+
+        return errors
+
+    def _validate_routing_when_dict(self, when_dict, answer_ids_with_parent_id, amount_matched):
+        errors = []
+
+        if when_dict.get('value'):
+            errors.extend(self._validate_routing_when_dict_value(when_dict, answer_ids_with_parent_id, amount_matched))
+
+        elif when_dict.get('values'):
+            errors.extend(self._validate_routing_when_dict_values(when_dict, answer_ids_with_parent_id, amount_matched))
+
+        return errors
+
+    @staticmethod
+    def _validate_routing_when_dict_value(when_dict, answer_ids_with_parent_id, amount_matched):
+        errors = []
+        if isinstance(when_dict.get('value'), int) is not True:
+            for block in answer_ids_with_parent_id:
+                if 'options' in answer_ids_with_parent_id[block]['answer']:
+                    for answer_block in answer_ids_with_parent_id[block]['answer'].get('options'):
+                        if when_dict.get('value') == answer_block.get('value'):
+                            amount_matched += 1
+            if amount_matched == 0:
+                errors.append(Validator._error_message(
+                    f'Answer option and routing rule values mismatch, missing answer value: '
+                    '{rule_value}'.format(rule_value=when_dict.get('value'))))
+        return errors
+
+    @staticmethod
+    def _validate_routing_when_dict_values(when_dict, answer_ids_with_parent_id, amount_matched):
+        errors = []
+        for rule_value in when_dict.get('values'):
+            if isinstance(rule_value, int) is not True:
+                for block in answer_ids_with_parent_id:
+                    if 'options' in answer_ids_with_parent_id[block]['answer']:
+                        for answer_block in answer_ids_with_parent_id[block]['answer'].get('options'):
+                            if rule_value == answer_block.get('value'):
+                                amount_matched += 1
+                if amount_matched == 0:
+                    errors.append(Validator._error_message(
+                        f'Answer option and routing rule values mismatch, missing answer value: '
+                        f'{rule_value}'))
+        return errors
```

Comment:
This needs refactoring - there is scope to make it significantly less complex.

1. You could make a generic function that takes a value, then pass each item from a `values` iterable as well as single `value` items to it, rather than having two very similar functions.
1. A  few of the if statements would be unnecessary if default values were used in the `.get` methods
1. There is a mixture of f-string and old style `.format` - f-strings are preferred
1. Do these need to be static methods?
1. There is a lot of nesting - a flat strucutre is generally better in terms of readability - is there a way to flatten this down?

Reply:
I think it's all addressed now, to a degree(except @staticmethod)